File: 1206.design-skiplist.py
import random

#
# @lc app=leetcode id=1206 lang=python3
#
# [1206] Design Skiplist
#

# @lc code=start
class Skiplist:

    MAX_LVL = 16
    P = 0.5

    class Node:
        def __init__(self, val: int, max_lvl: int):
            self.val = val
            self.next = [None] * max_lvl

    def __init__(self):
        self.head = self.Node(float('-inf'), self.MAX_LVL)

    def _random_level(self) -> int:
        lvl = 1
        while random.random() < self.P and lvl < self.MAX_LVL:
            lvl += 1
        return lvl

    def search(self, target: int) -> bool:
        cur = self.head
        for lvl in range(self.MAX_LVL - 1, -1, -1):
            while cur.next[lvl] is not None and cur.next[lvl].val < target:
                cur = cur.next[lvl]
        return cur.next[0] is not None and cur.next[0].val == target

    def add(self, num: int) -> None:
        lvl = self._random_level()
        new_node = self.Node(num, self.MAX_LVL)
        cur = self.head
        for i in range(self.MAX_LVL - 1, -1, -1):
            while cur.next[i] is not None and cur.next[i].val < num:
                cur = cur.next[i]
            if i < lvl:
                new_node.next[i] = cur.next[i]
                cur.next[i] = new_node

    def erase(self, num: int) -> bool:
        cur = self.head
        update = [None] * self.MAX_LVL
        for lvl in range(self.MAX_LVL - 1, -1, -1):
            while cur.next[lvl] is not None and cur.next[lvl].val < num:
                cur = cur.next[lvl]
            update[lvl] = cur
        if update[0].next[0] is None or update[0].next[0].val != num:
            return False
        node = update[0].next[0]
        for lvl in range(self.MAX_LVL):
            if update[lvl].next[lvl] == node:
                update[lvl].next[lvl] = node.next[lvl]
        return True


# Your Skiplist object will be instantiated and called as such:
# obj = Skiplist()
# param_1 = obj.search(target)
# obj.add(num)
# param_3 = obj.erase(num)
# @lc code=end

File: 1206.design-skiplist.cpp
#
# @lc app=leetcode id=1206 lang=cpp
#
# [1206] Design Skiplist
#
#include <vector>
#include <random>
#include <iostream>
using namespace std;
// @lc code=start
class Skiplist {
private:
    struct Node {
        int val;
        vector<Node*> forward;
        Node(int v, int level) : val(v), forward(level, nullptr) {}
    };
    const int MAX_LEVEL = 16; // enough for up to ~65k elements
    Node* head;
    int curMaxLevel;
    std::mt19937 rng;
    std::uniform_real_distribution<double> dist;
    
    int randomLevel() {
        // returns a random level between 1 and MAX_LEVEL inclusive
        // geometric distribution with p=0.5
        int lvl = 1;
        while ((dist(rng) < 0.5) && lvl < MAX_LEVEL)
            lvl++;
        return lvl;
    }
public:
    Skiplist() : curMaxLevel(1), rng(std::random_device{}()), dist(0.0, 1.0) {
        head = new Node(-1, MAX_LEVEL); // dummy head with value -1
    }
    
    ~Skiplist() {
        // clean up all nodes
        Node* curr = head->forward[0];
        while (curr) {
            Node* tmp = curr->forward[0];
            delete curr;
            curr = tmp;
        }
        delete head;
    }
    
    bool search(int target) {
        Node* cur = head;
        for (int i = curMaxLevel - 1; i >= 0; --i) {
            while (cur->forward[i] && cur->forward[i]->val < target)                cur = cur->forward[i];        }        cur = cur->forward[0];        return cur && cur->val == target;    }        void add(int num) {        vector<Node*> update(MAX_LEVEL, nullptr);        Node* cur = head;        // find predecessors at each layer        for (int i = curMaxLevel - 1; i >= 0; --i) {            while (cur->forward[i] && cur->forward[i]->val < num)                cur = cur->forward[i];            update[i] = cur;	}	// generate random level for new nod	int newLevel = randomLevel();	if(newLevel > curMaxLeve	{		for(int i=curMaxLeve;i<newLeve;i++)			update[i]=head		curMaxLeve=newLeve	}	Node* newNode=new Nod(numnewLeve	for(int i=0;i<newLeve;i++)	{		newNode->forwar[i]=update[i]->forwar[		update[		update[	}	}	bool erase(int num){	vector<Node*> updat(MAX_LEVE nullptr)	Node* cu=head	for(int i=curMaxLeve-1;i>=0;i--{	while(cu->forwar[	cu=cu->forwar[	upt[	upt[	upt[	upt[	upt[	upt[	upt[	upt[	upt[	upt[	upt[	upt[	upt[	upt[	t}	t}	t}	t}	t}	t}	t}	t}	t}	t}	t}	t}}
t}